diff --git a/hggit/hg2git.py b/hggit/hg2git.py
index ef27688..3103018 100644
--- a/hggit/hg2git.py
+++ b/hggit/hg2git.py
@@ -192,10 +192,12 @@ class IncrementalChangesetExporter(object):
         # The only reliable way to get the full set of changes is by looking at
         # the full manifest. And, the easy way to compare two manifests is
         # localrepo.status().
-        modified, added, removed = self._hg.status(self._ctx, newctx)[0:3]
+        status = self._hg.status(self._ctx, newctx)
+
+        modified = status.modified
+        added = status.added
+        removed = status.removed
 
-        # We track which directories/trees have modified in this update and we
-        # only export those.
         dirty_trees = set()
 
         subadded, subremoved = [], []
